### PR TITLE
Speed up Out Of Scope lines

### PR DESCRIPTION
**Pull request by jasonLaster**
Associated Issue: #3180

### Summary of Changes

This refactors out of scope lines a bit to support getting a list of in scope lines that can be used by the editor to check if a token is in scope on hover.

### Test Plan

Jest tests

### CODE REVIEW

File: src/components/Editor/index.js [hunk 1]
```diff
@@ -155,8 +154,10 @@ class Editor extends PureComponent {
       this.showSourceText(selectedSource, selectedLocation);
     }
 
-    if (this.props.outOfScopeLocations !== nextProps.outOfScopeLocations) {
-      clearLineClass(this.editor.codeMirror, "out-of-scope");
+    if (this.props.linesInScope !== nextProps.linesInScope) {
+      this.editor.codeMirror.operation(() => {
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: `this.editor.codeMirror.operation` is a huge perf win. With that said, i'll be moving the operation inside `clearLineClass` soon :)

File: src/components/Editor/index.js [hunk 2]
```diff
@@ -788,6 +795,23 @@ class Editor extends PureComponent {
     });
   }
 
+  renderInScopeLines() {
+    const { linesInScope } = this.props;
+    if (
+      !isEnabled("highlightScopeLines") ||
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: i'm re-adding a feature flag because i'm not happy with the cases we're missing like blocks and comments. Seeing those issues makes me think we should wait and re-launch highlighting when it's further along. 

With that said, we still get the onMouseOver logic so fewer accidental popups

File: src/selectors/linesInScope.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// import { linesInScope } from "../utils/scopes";
+import { getOutOfScopeLocations, getSelectedSource } from "../selectors";
```
Comment (jasonLaster) [inline src/selectors/linesInScope.js]: this is the first selector that spans stores (sources, ast) with that in mind I created a new selectors directory. I'm not sure if that makes sense. What do others think?

File: src/actions/ast.js [hunk 4]
```diff
@@ -1,33 +1,34 @@
...
-    if (hasSymbols(getState(), source)) {
+    const sourceRecord = getSource(getState(), sourceId);
+    if (!sourceRecord) {
       return;
     }
 
-    const sourceText = getSourceText(getState(), source.id);
-    if (!sourceText) {
+    const source = sourceRecord.toJS();
+    if (!source.text || hasSymbols(getState(), source)) {
```
Comment (jasonLaster) [inline src/actions/ast.js]: some small `sourceText => source` migrations

File: src/actions/tests/ast.js [hunk 5]
```diff
@@ -72,32 +75,33 @@ describe("ast", () => {
...
   describe("getOutOfScopeLocations", () => {
-    it("simple", async () => {
+    it("with selected line", async () => {
       const { dispatch, getState } = createStore(threadClient);
-      const base = makeSource("base.js");
-      await dispatch(actions.newSource(base));
-      await dispatch(actions.loadSourceText({ id: "base.js" }));
-      await dispatch(actions.selectSource("base.js", { line: 1 }));
-
-      await dispatch(actions.setOutOfScopeLocations());
```
Comment (jasonLaster) [inline src/actions/tests/ast.js]: yeah - so this is coooool. Turns out many of the dispatches are unnecessary.

File: src/components/Editor/Editor.css [hunk 6]
```diff
@@ -22,8 +22,12 @@
...
 }
 
-.out-of-scope .CodeMirror-line, .out-of-scope .CodeMirror-linenumber {
-  opacity: 0.8;
+.paused .in-scope .CodeMirror-line, .paused .in-scope .CodeMirror-linenumber {
+  opacity: 1;
+}
+
+.paused .CodeMirror-line, .paused .CodeMirror-linenumber {
+  opacity: 0.7;
```
Comment (jasonLaster) [inline src/components/Editor/Editor.css]: we're now tracking in-scope and not out-of-scope. 

some reasons:
1. it's easier to get the lines in scope than all the lines that are out of scope
2. it should be faster because there are fewer innies than outies.

File: src/reducers/ast.js [hunk 7]
```diff
@@ -79,6 +79,10 @@ function update(
       });
     }
 
+    case "RESUMED": {
+      return state.set("outOfScopeLocations", null);
+    }
+
```
Comment (jasonLaster) [inline src/reducers/ast.js]: this fixes #3123

File: src/selectors/linesInScope.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+import range from "lodash/range";
```
Comment (codehag) [inline src/selectors/linesInScope.js]: would it make sense to do this?

```javascript
import { range, flatMap, uniq, without } from "lodash";
```

File: src/selectors/linesInScope.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import flatMap from "lodash/flatMap";
+import uniq from "lodash/uniq";
+import without from "lodash/without";
+
+function getLinesOutOfScope(outOfScopeLocations: AstLocation[]) {
+  if (!outOfScopeLocations) {
```
Comment (codehag) [inline src/selectors/linesInScope.js]: if this is an array of AstLocations, I think this will always be true... or maybe it isnt an array?

if it is an array we probably need to do 
```javascript
if (!outOfScopeLocations.length) {
  ...
}
```

File: src/selectors/linesInScope.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+import { getOutOfScopeLocations, getSelectedSource } from "../selectors";
+
+import range from "lodash/range";
+import flatMap from "lodash/flatMap";
+import uniq from "lodash/uniq";
+import without from "lodash/without";
+
+function getLinesOutOfScope(outOfScopeLocations: AstLocation[]) {
```
Comment (codehag) [inline src/selectors/linesInScope.js]: `getLinesOutOfScope` sounds like we are taking lines out of a scope, but I think this is getting the out of scope lines. maybe `getOutOfScopeLines` will be clearer here,

File: src/selectors/linesInScope.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  if (!outOfScopeLocations) {
+    return null;
+  }
+
+  return uniq(
+    flatMap(outOfScopeLocations, location =>
+      range(location.start.line, location.end.line)
```
Comment (codehag) [inline src/selectors/linesInScope.js]: nice

File: src/selectors/linesInScope.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    )
+  );
+}
+
+export default function getLinesInScope(state: OuterState) {
```
Comment (codehag) [inline src/selectors/linesInScope.js]: same comment here about naming: proposing `getInScopeLines`

File: src/selectors/linesInScope.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
+import { getOutOfScopeLocations, getSelectedSource } from "../selectors";
+
+import range from "lodash/range";
+import flatMap from "lodash/flatMap";
+import uniq from "lodash/uniq";
+import without from "lodash/without";
+
+function getLinesOutOfScope(outOfScopeLocations: AstLocation[]) {
+  if (!outOfScopeLocations) {
+    return null;
```
Comment (codehag) [inline src/selectors/linesInScope.js]: this is returning a different type; maybe it should be an empty array?

File: src/selectors/linesInScope.js [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+  const linesOutOfScope = getLinesOutOfScope(
+    outOfScopeLocations,
+    source.toJS()
+  );
+
+  const sourceNumLines = source.get("text").split("\n").length;
+  const sourceLines = range(1, sourceNumLines + 1);
+
+  if (!linesOutOfScope) {
```
Comment (codehag) [inline src/selectors/linesInScope.js]: the logic here would be clearer if this was checking for an empty array

File: src/actions/tests/ast.js (hunk 5 above)
Comment (codehag) [inline src/actions/tests/ast.js]: awesome

File: src/selectors/linesInScope.js (hunk 8 above)
Comment (jasonLaster) [inline src/selectors/linesInScope.js]: sadly... no because then we pull lodash into the bundle... I suppose if we made it an external then we could exclude it from the bundle and expect it to be available in m-c... If we like this i can file a separate issue for it.

File: src/selectors/linesInScope.js (hunk 10 above)
Comment (jasonLaster) [inline src/selectors/linesInScope.js]: agreed :)

File: src/selectors/linesInScope.js (hunk 9 above)
Comment (jasonLaster) [inline src/selectors/linesInScope.js]: I can do both. when it is not available it's nully.

File: src/selectors/linesInScope.js (hunk 13 above)
Comment (jasonLaster) [inline src/selectors/linesInScope.js]: ... not sure. i kinda like having explicit empty states to distinguish between nothing and we happen to have no out of scope locs. ill think about it

File: src/selectors/linesInScope.js (hunk 8 above)
Comment (codehag) [inline src/selectors/linesInScope.js]: :+1:
